File: app/src/main/java/com/theappsolutions/boilerplate/util/data_utils/RandomUtils.java
package com.theappsolutions.boilerplate.util.data_utils;

import com.annimon.stream.Objects;
import com.annimon.stream.Stream;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import java.util.Set;

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 */
public class RandomUtils {

    private static final String upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String lower = upper.toLowerCase(Locale.ROOT);
    private static final String digits = "0123456789";
    private static final String alphanum = upper + lower + digits;
    private final Random random;
    private final char[] symbols;
    private final char[] buf;

    public RandomUtils(int length, Random random, String symbols) {
        if (length < 1) throw new IllegalArgumentException();
        if (symbols.length() < 2) throw new IllegalArgumentException();
        this.random = Objects.requireNonNull(random);
        this.symbols = symbols.toCharArray();
        this.buf = new char[length];
    }

    /**
     * Create an alphanumeric string generator.
     */
    public RandomUtils(int length, Random random) {
        this(length, random, alphanum);
    }

    /**
     * Create an alphanumeric strings from a secure generator.
     */
    public RandomUtils(int length) {
        this(length, new SecureRandom());
    }

    /**
     * Create session identifiers.
     */
    public RandomUtils() {
        this(21);
    }

    /**
     * Generate a random string.
     */
    public String nextString() {
        for (int idx = 0; idx < buf.length; ++idx)
            buf[idx] = symbols[random.nextInt(symbols.length)];
        return new String(buf);
    }

    public static <T> List<T> generateRandList(List<T> src, int quantity) {
        if (src.size() <= quantity) {
            return src;
        }
        Set<Integer> indexes = new HashSet<>();
        Random random = new Random();
        while (indexes.size() < quantity) {
            int newRand = random.nextInt(src.size());
            indexes.add(newRand);
        }
        List<T> dest = new ArrayList<>();
        Stream.of(indexes).forEach(index -> dest.add(src.get(index)));
        return dest;
    }
}

File: app/src/main/java/com/theappsolutions/boilerplate/util/data/RandomUtils.kt
package com.theappsolutions.boilerplate.util.data


import java.security.SecureRandom
import java.util.ArrayList
import java.util.HashSet
import java.util.Locale
import java.util.Random

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 */
class RandomUtils {
    private val random: Random
    private val symbols: CharArray
    private val buf: CharArray

    @JvmOverloads
    constructor(length: Int = 21, random: Random = SecureRandom(), symbols: String = alphanum) {
        if (length < 1) throw IllegalArgumentException()
        if (symbols.length < 2) throw IllegalArgumentException()
        this.random = random
        this.symbols = symbols.toCharArray()
        this.buf = CharArray(length)
    }


    /**
     * Generate a random string.
     */
    fun nextString(): String {
        buf.indices.forEach { idx -> buf[idx] = symbols[random.nextInt(symbols.size)] }
        return String(buf)
    }

    companion object {
        private val upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        private val lower = upper.toLowerCase(Locale.ROOT)
        private val digits = "0123456789"
        private val alphanum = upper + lower + digits

        fun <T> generateRandList(src: List<T>, quantity: Int): List<T> {
            if (src.size <= quantity) {
                return src
            }
            val indexes = HashSet<Int>()
            val random = Random()
            while (indexes.size < quantity) {
                val newRand = random.nextInt(src.size)
                indexes.add(newRand)
            }
            val dest = ArrayList<T>()
            indexes.forEach { index -> dest.add(src[index]) }
            return dest
        }
    }
}
/**
 * Create an alphanumeric string generator.
 */
/**
 * Create an alphanumeric strings from a secure generator.
 */
/**
 * Create session identifiers.
 */
